File: Backend/TaxSummary.Domain/ValueObjects/TaxHierarchy.cs
using System.Text.RegularExpressions;
using TaxSummary.Domain.Common;

namespace TaxSummary.Domain.ValueObjects;

/// <summary>
/// Represents the 3-tier organizational tax hierarchy in the Iranian Tax System:
/// Level 1: Office (اداره امور مالیاتی) - e.g. 160200 (Presided by رئیس امور)
/// Level 2: Audit Group (گروه رسیدگی مالیاتی) - e.g. 160210 (Presided by رئیس گروه)
/// Level 3: Tax Unit (واحد مالیاتی) - e.g. 160211 (Presided by کارشناس ارشد مالیاتی)
/// </summary>
public class TaxHierarchy : ValueObject
{
    private static readonly Regex SixDigitRegex = new(@"^\d{6}$", RegexOptions.Compiled);

    public string TaxUnitCode { get; private set; } = string.Empty;
    public string GroupCode { get; private set; } = string.Empty;
    public string OfficeCode { get; private set; } = string.Empty;

    public int? GroupIndex { get; private set; }
    public int? UnitIndex { get; private set; }

    /// <summary>
    /// Indicates whether the tax unit code decomposed into a standard 6-digit hierarchy
    /// </summary>
    public bool IsValid => !string.IsNullOrWhiteSpace(TaxUnitCode) && SixDigitRegex.IsMatch(TaxUnitCode);

    private TaxHierarchy() { }

    private TaxHierarchy(string taxUnitCode, string groupCode, string officeCode, int? groupIndex = null, int? unitIndex = null)
    {
        TaxUnitCode = taxUnitCode;
        GroupCode = groupCode;
        OfficeCode = officeCode;
        GroupIndex = groupIndex;
        UnitIndex = unitIndex;
    }

    /// <summary>
    /// Decomposes any tax unit code into its hierarchical Office (Level 1), Group (Level 2), and Unit (Level 3)
    /// </summary>
    public static TaxHierarchy FromTaxUnitCode(string? code) => Decompose(code);

    public static TaxHierarchy Decompose(string? code)
    {
        if (string.IsNullOrWhiteSpace(code))
            return new TaxHierarchy(string.Empty, string.Empty, string.Empty);

        var normalized = NormalizeDigits(code.Trim().Replace("-", "").Replace(" ", ""));

        if (SixDigitRegex.IsMatch(normalized))
        {
            var officePrefix = normalized.Substring(0, 4);
            var groupDigit = normalized[4];
            var unitDigit = normalized[5];

            var officeCode = $"{officePrefix}00";
            var groupCode = $"{officePrefix}{groupDigit}0";
            var unitCode = normalized;

            int.TryParse(groupDigit.ToString(), out var gIndex);
            int.TryParse(unitDigit.ToString(), out var uIndex);

            return new TaxHierarchy(unitCode, groupCode, officeCode, gIndex, uIndex);
        }

        if (normalized.Length == 5 && int.TryParse(normalized, out _))
        {
            var officePrefix = normalized.Substring(0, 4);
            var groupDigit = normalized[4];
            var officeCode = $"{officePrefix}00";
            var groupCode = $"{normalized}0";
            var unitCode = $"{normalized}0";

            int.TryParse(groupDigit.ToString(), out var gIndex);

            return new TaxHierarchy(unitCode, groupCode, officeCode, gIndex, 0);
        }

        if (normalized.Length == 4 && int.TryParse(normalized, out _))
        {
            var officeCode = $"{normalized}00";
            var groupCode = $"{normalized}00";
            var unitCode = $"{normalized}00";

            return new TaxHierarchy(unitCode, groupCode, officeCode, 0, 0);
        }

        // Fallback for non-standard or custom codes
        return new TaxHierarchy(normalized, normalized, normalized);
    }

    /// <summary>
    /// Checks whether an assigned user code (e.g. 160200 for Office, 160210 for Group, 160211 for Unit) covers this hierarchy.
    /// </summary>
    public bool IsCoveredBy(string? assignedCode)
    {
        if (string.IsNullOrWhiteSpace(assignedCode)) return false;

        var clean = NormalizeDigits(assignedCode.Trim().Replace("-", "").Replace(" ", ""));

        if (clean.Equals(TaxUnitCode, StringComparison.OrdinalIgnoreCase) ||
            clean.Equals(GroupCode, StringComparison.OrdinalIgnoreCase) ||
            clean.Equals(OfficeCode, StringComparison.OrdinalIgnoreCase))
        {
            return true;
        }

        // Office level match (e.g., assigned 160200 or 1602 covers any 1602xx)
        if (clean.Length == 6 && clean.EndsWith("00"))
        {
            var officePrefix = clean.Substring(0, 4);
            return TaxUnitCode.StartsWith(officePrefix, StringComparison.OrdinalIgnoreCase);
        }

        if (clean.Length == 4)
        {
            return TaxUnitCode.StartsWith(clean, StringComparison.OrdinalIgnoreCase);
        }

        // Group level match (e.g., assigned 160210 covers 160211 - 160219)
        if (clean.Length == 6 && clean.EndsWith("0") && !clean.EndsWith("00"))
        {
            var groupPrefix = clean.Substring(0, 5);
            return TaxUnitCode.StartsWith(groupPrefix, StringComparison.OrdinalIgnoreCase);
        }

        return false;
    }

    /// <summary>
    /// Converts Persian and Arabic unicode digits to ASCII standard digits
    /// </summary>
    public static string NormalizeDigits(string input)
    {
        if (string.IsNullOrEmpty(input))
            return string.Empty;

        return input
            .Replace('۰', '0').Replace('۱', '1').Replace('۲', '2').Replace('۳', '3').Replace('۴', '4')
            .Replace('۵', '5').Replace('۶', '6').Replace('۷', '7').Replace('۸', '8').Replace('۹', '9')
            .Replace('٠', '0').Replace('١', '1').Replace('٢', '2').Replace('٣', '3').Replace('٤', '4')
            .Replace('٥', '5').Replace('٦', '6').Replace('٧', '7').Replace('٨', '8').Replace('٩', '9');
    }

    protected override IEnumerable<object> GetEqualityComponents()
    {
        yield return TaxUnitCode;
        yield return GroupCode;
        yield return OfficeCode;
    }
}

File: frontend/lib/taxHierarchy.ts
import { toPersianDigits } from './jalali'

export interface TaxHierarchyInfo {
  taxUnitCode: string
  groupCode: string
  officeCode: string
  officeName: string
  groupName: string
  unitName: string
  isValid: boolean
}

// Known human-readable names for Khuzestan tax offices (1601-1622 are Ahvaz tax offices)
export const KNOWN_OFFICES: Record<string, string> = {
  '160100': 'اداره امور مالیاتی ۱ اهواز',
  '160200': 'اداره امور مالیاتی ۲ اهواز',
  '160300': 'اداره امور مالیاتی ۳ اهواز',
  '160400': 'اداره امور مالیاتی ۴ اهواز',
  '160500': 'اداره امور مالیاتی ۵ اهواز',
  '160600': 'اداره امور مالیاتی ۶ اهواز',
  '160700': 'اداره امور مالیاتی ۷ اهواز',
  '160800': 'اداره امور مالیاتی ۸ اهواز',
  '160900': 'اداره امور مالیاتی ۹ اهواز',
  '161000': 'اداره امور مالیاتی ۱۰ اهواز',
  '161100': 'اداره امور مالیاتی ۱۱ اهواز',
  '161200': 'اداره امور مالیاتی ۱۲ اهواز',
  '161300': 'اداره امور مالیاتی ۱۳ اهواز',
  '161400': 'اداره امور مالیاتی ۱۴ اهواز',
  '161500': 'اداره امور مالیاتی ۱۵ اهواز',
  '161600': 'اداره امور مالیاتی ۱۶ اهواز',
  '161700': 'اداره امور مالیاتی ۱۷ اهواز',
  '161800': 'اداره امور مالیاتی ۱۸ اهواز',
  '161900': 'اداره امور مالیاتی ۱۹ اهواز',
  '162000': 'اداره امور مالیاتی ۲۰ اهواز',
  '162100': 'اداره امور مالیاتی ۲۱ اهواز',
  '162200': 'اداره امور مالیاتی ۲۲ اهواز',
}

/**
 * Decomposes an Iranian 6-digit tax unit code into its 3 organizational tiers:
 * Level 1: Office (اداره امور مالیاتی) -> e.g., 160200
 * Level 2: Audit Group (گروه رسیدگی مالیاتی) -> e.g., 160210
 * Level 3: Tax Unit (واحد مالیاتی) -> e.g., 160211
 */
export function decomposeTaxUnitCode(code?: string | null): TaxHierarchyInfo {
  if (!code) {
    return {
      taxUnitCode: '',
      groupCode: '',
      officeCode: '',
      officeName: '',
      groupName: '',
      unitName: '',
      isValid: false,
    }
  }

  // Normalize Persian/Arabic digits to ASCII standard
  const normalized = code
    .replace(/[۰-۹]/g, (d) => '۰۱۲۳۴۵۶۷۸۹'.indexOf(d).toString())
    .replace(/[٠-٩]/g, (d) => '٠١٢٣٤٥٦٧٨٩'.indexOf(d).toString())
    .replace(/[^0-9]/g, '')
    .trim()

  if (normalized.length === 6) {
    const officePrefix = normalized.substring(0, 4)
    const groupDigit = normalized[4]
    const unitDigit = normalized[5]

    const officeCode = `${officePrefix}00`
    const groupCode = `${officePrefix}${groupDigit}0`
    const taxUnitCode = normalized

    let officeName = KNOWN_OFFICES[officeCode]
    if (!officeName) {
      if (officePrefix.startsWith('16') && officePrefix.length === 4) {
        const num = parseInt(officePrefix.substring(2), 10)
        if (!isNaN(num)) {
          officeName = `اداره امور مالیاتی ${toPersianDigits(num)} اهواز`
        }
      }
      if (!officeName) {
        officeName = `اداره امور مالیاتی ${toPersianDigits(officePrefix)}`
      }
    }
    const groupName = `گروه رسیدگی ${toPersianDigits(groupDigit)}`
    const unitName = `واحد مالیاتی ${toPersianDigits(unitDigit)}`

    return {
      taxUnitCode,
      groupCode,
      officeCode,
      officeName,
      groupName,
      unitName,
      isValid: true,
    }
  }

  return {
    taxUnitCode: normalized,
    groupCode: normalized,
    officeCode: normalized,
    officeName: `اداره امور مالیاتی ${normalized}`,
    groupName: '',
    unitName: '',
    isValid: false,
  }
}

/**
 * Checks whether a user's assigned offices/codes cover the target tax unit hierarchy.
 */
export function isUserAuthorizedForUnit(
  userRole?: string,
  userOffices?: { code: string; name?: string }[],
  taxUnitCode?: string,
  serviceUnit?: string
): boolean {
  if (!userRole) return false
  if (userRole.toLowerCase() === 'admin') return true
  if (!taxUnitCode) return false

  const hierarchy = decomposeTaxUnitCode(taxUnitCode)
  const codes: string[] = []

  if (userOffices && userOffices.length > 0) {
    userOffices.forEach((o) => {
      if (o.code) codes.push(o.code.trim())
    })
  }

  if (serviceUnit && serviceUnit.trim()) {
    codes.push(serviceUnit.trim())
  }

  if (codes.length === 0) return false

  return codes.some((code) => {
    if (!code) return false

    // Exact match on unit, group, or office
    if (code === hierarchy.taxUnitCode || code === hierarchy.groupCode || code === hierarchy.officeCode) {
      return true
    }

    // Office level code (e.g. 160200 or 1602) covers any 1602xx
    if (code.length === 6 && code.endsWith('00')) {
      return hierarchy.taxUnitCode.startsWith(code.substring(0, 4))
    }
    if (code.length === 4) {
      return hierarchy.taxUnitCode.startsWith(code)
    }

    // Group level code (e.g. 160210) covers 160211 - 160219
    if (code.length === 6 && code.endsWith('0') && !code.endsWith('00')) {
      return hierarchy.taxUnitCode.startsWith(code.substring(0, 5))
    }

    return false
  })
}

/**
 * Checks whether a user has permission to verify/approve a specific stage in the refund workflow
 * according to their role and organizational tier.
 */
export function canUserVerifyStage(
  userRole?: string,
  userOffices?: { code: string; name?: string }[],
  serviceUnit?: string,
  targetStatus?: number | string,
  taxUnitCode?: string
): boolean {
  if (!userRole) return false
  const role = userRole.toLowerCase()
  if (role === 'admin') return true

  // Must have hierarchy access to the tax unit code
  if (!isUserAuthorizedForUnit(userRole, userOffices, taxUnitCode, serviceUnit)) {
    return false
  }

  const statusNum = typeof targetStatus === 'number' ? targetStatus : parseInt(targetStatus as string, 10)

  // Status mapping:
  // 1 = Audited (Auditor / Expert, GroupHead, OfficeHead, Admin)
  // 2 = GroupHeadApproved (GroupHead, OfficeHead, Admin)
  // 3 = AdministrationHeadApproved (OfficeHead, Admin)
  switch (statusNum) {
    case 1:
      return role === 'expert' || role === 'grouphead' || role === 'officehead' || role === 'manager'
    case 2:
      return role === 'grouphead' || role === 'officehead'
    case 3:
      return role === 'officehead'
    default:
      return true
  }
}

